app/org/sagebionetworks/bridge/models/schedules/ActivityScheduler.java

```diff
@@ -125,28 +141,43 @@ private LocalDateTime getExpiresOn(LocalDate localDate, LocalTime localTime) {
         return localDate.toLocalDateTime(localTime).plus(schedule.getExpires());
     }
 
-    protected DateTime getFirstEventDateTime(ScheduleContext context, String eventIdsString) {
-        DateTime eventDateTime = null;
+    /**
+     * Helper method used to get specified event date times from the schedule context.
+     *
+     * @param context
+     *         schedule context, which contains the event map
+     * @param eventIdsString
+     *         event ID string as specified by the schedule; this is a comma-delimited list
+     * @param getAll
+     *         true to get all events; false to just get any one event (generally useful for persistent activities)
+     * @return list of one or all specified event date-times
+     */
+    protected List<DateTime> getEventDateTimes(ScheduleContext context, String eventIdsString, boolean getAll) {
+        List<DateTime> eventDateTimeList = new ArrayList<>();
         if (eventIdsString != null) {
             Iterable<String> eventIds = Schedule.EVENT_ID_SPLITTER.split(eventIdsString.trim());
             for (String thisEventId : eventIds) {
                 if (context.getEvent(thisEventId) != null) {
-                    eventDateTime = context.getEvent(thisEventId);
-                    break;
+                    eventDateTimeList.add(context.getEvent(thisEventId));
+
+                    if (!getAll) {
+                        // We only wanted one event, and we found it, so break.
+                        break;
+                    }
```

Comment:
A common event list is something like "enrollment,actionFinished" (not the exact syntax). For a one time event, this would schedule an initial one on enrollment, and then another one on the actionFinished event. Will this still work at enrollment. It seems like yes, there's no actionFinished event, but once actionFinished event occurs, it still works? I presume it will keep scheduling an enrollment event, finding it is finished, and not returning it?

Reply:
I believe that ActivitySchedulerTest.willSelectFirstEventIdWithARecord() https://github.com/Sage-Bionetworks/BridgePF/blob/develop/test/org/sagebionetworks/bridge/models/schedules/ActivitySchedulerTest.java#L243

This test tests the scenario you describe. If both survey:event and enrollment are in the events map, it'll schedule the event off survey:event. If there's no survey:event, it'll schedule the event off enrollment. Note that one-time schedules always return only 1 event, because of the trim logic.

Note that there's no functional change for one-time or persistent schedules. This change only affects recurring schedules.